packages/grid/x-data-grid/src/hooks/features/pagination/useGridPagination.ts

```diff
@@ -1,41 +1,81 @@
-} from './useGridPageSize';
 import { GridPrivateApiCommunity } from '../../../models/api/gridApiCommunity';
 import { DataGridProcessedProps } from '../../../models/props/DataGridProps';
-import { useGridPage, getPageCount } from './useGridPage';
 import { GridStateInitializer } from '../../utils/useGridInitializeState';
+import { GridPaginationApi, GridPaginationState } from './gridPaginationInterfaces';
+import { GridEventListener } from '../../../models/events';
+import { GridPaginationModel } from '../../../models/gridPaginationProps';
+import { gridVisibleTopLevelRowCountSelector } from '../filter';
+import { gridDensityFactorSelector } from '../density';
+import {
+  useGridLogger,
+  useGridSelector,
+  useGridApiMethod,
+  useGridApiEventHandler,
+} from '../../utils';
+import { GridPipeProcessor, useGridRegisterPipeProcessor } from '../../core/pipeProcessing';
+import { gridPaginationModelSelector } from './gridPaginationSelector';
+import { calculatePinnedRowsHeight } from '../rows/gridRowsUtils';
+import {
+  getPageCount,
+  noRowCountInServerMode,
+  defaultPageSize,
+  throwIfPageSizeExceedsTheLimit,
+  getDefaultGridPaginationModel,
+  getValidPage,
+} from './gridPaginationUtils';
 
 export const paginationStateInitializer: GridStateInitializer<
   Pick<
     DataGridProcessedProps,
-    'page' | 'pageSize' | 'rowCount' | 'initialState' | 'autoPageSize' | 'signature'
+    'paginationModel' | 'rowCount' | 'initialState' | 'autoPageSize' | 'signature'
   >
 > = (state, props) => {
-  let pageSize: number;
-  if (props.pageSize != null) {
-    pageSize = props.pageSize;
-  } else if (props.initialState?.pagination?.pageSize != null) {
-    pageSize = props.initialState.pagination.pageSize;
-    throwIfPageSizeExceedsTheLimit(pageSize, props.signature);
-  } else {
-    pageSize = defaultPageSize(props.autoPageSize);
-  }
+  const paginationModel = {
+    ...getDefaultGridPaginationModel(props.autoPageSize),
+    ...(props.paginationModel ?? props.initialState?.pagination?.paginationModel),
+  };
+
+  throwIfPageSizeExceedsTheLimit(paginationModel.pageSize, props.signature);
 
   return {
     ...state,
     pagination: {
-      pageSize,
-      page: props.page ?? props.initialState?.pagination?.page ?? 0,
-      pageCount: getPageCount(props.rowCount ?? 0, pageSize),
+      paginationModel,
+      pageCount: getPageCount(props.rowCount ?? 0, paginationModel.pageSize),
       rowCount: props.rowCount ?? 0,
     },
   };
 };
 
+const mergeStateWithPaginationModel =
+  (
+    rowCount: number,
+    signature: DataGridProcessedProps['signature'],
+    paginationModelProp?: GridPaginationModel,
+  ) =>
+  (paginationState: GridPaginationState) => {
+    let paginationModel = paginationState.paginationModel;
+    const pageSize = paginationModelProp?.pageSize ?? paginationModel.pageSize;
+    const pageCount = getPageCount(rowCount, pageSize);
+
+    if (paginationModelProp && paginationModelProp !== paginationModel) {
```

Comment:
The possibility of users storing values on their side in separate state variables etc could surely be a use-case. 👍 

I wanted to go with shallow equality for the scalability purpose, but since I can't foresee too many new props being added to `paginationModel` soon, I'll refactor it.